csr_check_test.go

```diff
@@ -1992,6 +2355,177 @@ func TestAuthorizeServingRenewal(t *testing.T) {
+	defer server.Close()
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			var certPool *x509.CertPool
+			if len(tt.rootCerts) > 0 {
+				certPool = x509.NewCertPool()
+				for _, cert := range tt.rootCerts {
+					certPool.AddCert(cert)
+				}
+			}
+
+			nodes := &testNode{t: t, node: tt.node, err: tt.serverErr}
+
+			go respond(server)
+			serverCert, err := getServingCert(nodes, tt.nodeName, certPool)
+
+			if errString(err) != tt.wantErr {
+				t.Fatalf("got: %v, want: %s", err, tt.wantErr)
+			}
+			if err == nil && !serverCert.Equal(parseCert(t, serverCertGood)) {
+				t.Fatal("Expected server certificate match on success")
+			}
+		})
+	}
+}
+
+func TestRecentlyPendingCSRs(t *testing.T) {
+	approvedCSR := certificatesv1beta1.CertificateSigningRequest{
+		Status: certificatesv1beta1.CertificateSigningRequestStatus{
+			Conditions: []certificatesv1beta1.CertificateSigningRequestCondition{{
+				Type: certificatesv1beta1.CertificateApproved,
+			}},
+		},
+	}
+	pendingCSR := certificatesv1beta1.CertificateSigningRequest{}
+	pendingTime := baseTime.Add(time.Second)
+	pastApprovalTime := baseTime.Add(-maxPendingDelta)
+	preApprovalTime := baseTime.Add(10 * time.Second)
+
+	createdAt := func(time time.Time, csr certificatesv1beta1.CertificateSigningRequest) *certificatesv1beta1.CertificateSigningRequest {
+		csr.CreationTimestamp.Time = time
+		return csr.DeepCopy()
+	}
+
+	tests := []struct {
+		name          string
+		csrs          []*certificatesv1beta1.CertificateSigningRequest
+		expectPending int
+	}{
+		{
+			name:          "recently pending csr",
+			csrs:          []*certificatesv1beta1.CertificateSigningRequest{createdAt(pendingTime, pendingCSR)},
+			expectPending: 1,
+		},
+		{
+			name:          "recently approved csr",
+			csrs:          []*certificatesv1beta1.CertificateSigningRequest{createdAt(pendingTime, approvedCSR)},
+			expectPending: 0,
+		},
+		{
+			name:          "pending past approval time",
+			csrs:          []*certificatesv1beta1.CertificateSigningRequest{createdAt(pastApprovalTime, pendingCSR)},
+			expectPending: 0,
+		},
+		{
+			name:          "pending before approval time",
+			csrs:          []*certificatesv1beta1.CertificateSigningRequest{createdAt(preApprovalTime, pendingCSR)},
+			expectPending: 0,
+		},
+		{
+			name: "multiple different csrs",
+			csrs: []*certificatesv1beta1.CertificateSigningRequest{
+				createdAt(pendingTime, pendingCSR),
+				createdAt(pendingTime, pendingCSR),
+
+				createdAt(pendingTime, approvedCSR),
+				createdAt(preApprovalTime, approvedCSR),
+				createdAt(pastApprovalTime, approvedCSR),
+				createdAt(preApprovalTime, pendingCSR),
+				createdAt(pastApprovalTime, pendingCSR),
+			},
+			expectPending: 2,
+		},
+	}
+
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			if pending := recentlyPendingCSRs(tt.csrs); pending != tt.expectPending {
+				t.Errorf("Expected %v pending CSRs, got: %v", tt.expectPending, pending)
```

Comment:
Looking at the way these tests are structured, we could easily use a fake lister here to avoid having to move the logic in the main code flow